README.md
tapirlink
=========

Tapirlink made compatible with darwincore v2011

Below is the former contents of "README.txt"

ABOUT THIS SOFTWARE
===================

TapirLink is a data provider software compatible with the 
TAPIR protocol (http://www.tdwg.org/subgroups/tapir). 

"TAPIR specifies a standardised, stateless, HTTP transmittable, 
request and response protocol for accessing structured data that 
may be stored on any number of distributed databases of varied 
physical and logical structure."

Data provider software is one of the key components of TAPIR 
networks. All participants of a TAPIR network, ie. people or 
institutions that want to share data through TAPIR, need to 
install a data provider software. TapirLink allows data providers 
to map their local databases against one or more conceptual 
schemas (used as a data abstraction layer). It translates TAPIR 
search requests to the local query language and local database 
structure to return an XML response based on the output model 
requested.

ACKNOWLEDGEMENTS
================ 

TapirLink has been generously funded by the Biodiversity 
Information Standards, TDWG, with resources from the Gordon and 
Betty Moore Foundation. The Global Biodiversity Information 
Facility, GBIF, has also been a major supporter of the TAPIR 
initiative since the very beginning and also collaborated to 
test TapirLink.

This software was based on the DiGIR PHP provider, which was 
originally developed by Dave Vieglais from the Biodiversity 
Research Center, University of Kansas. Two other institutions 
participated in the DiGIR project: California Academy of Sciences 
and Museum of Vertebrate Zoology, Berkeley.

The KVP filter parsing functionality in TapirLink has been ported 
from PyWrapper code (http://www.pywrapper.org), whose main author 
is Markus Döring.

NOTES
=====

Although TapirLink is based on the DiGIR PHP provider, it is
NOT compatible with the DiGIR protocol. TapirLink is only 
compatible with the TAPIR protocol (version 1.0 of the 
protocol, as specified at: 
http://www.tdwg.org/activities/tapir/specification/

A single instance of TapirLink can provide access to multiple 
TAPIR resources, each one with its own address. After 
installing this software and adjusting your web server 
configuration, you may be tempted to guess that the access 
point of your service is something like:

http://example.net/tapirlink/tapir.php

However, this will only give you generic documentation about 
the service. Real interaction with the service can only happen 
through one of the end points. After configuring TapirLink, you 
will notice that each resource has a local id (or code). The 
local id must be appended to the previous URI to give you the 
corresponding address of the service, like: 

http://example.net/tapirlink/tapir.php/myres/

TapirLink can map conceptual schemas that either follow the 
DarwinCore patterns or the CNS configuration file patterns.

For specimen/observation data providers, TapirLink has been 
tested with:

* The original version of DarwinCore;
* The second generation of DarwinCore (with the curatorial and 
geospatial extensions);
* The official version of DarwinCore approved by TDWG
(DarwinCore terms);
* ABCD 2.06

For taxonomic data providers TapirLink has been tested with:

* TCS 1.01

TapirLink allows each resource to map one or more conceptual 
schemas, but it will only be able to serve instances of a single 
"class" or "entity". In other words, when mapping multiple 
conceptual schemas, each mapped concept will actually refer to an 
"attribute" or "property" of the same underlying class. In search 
responses, instances of that class will be bound to instances of 
the "indexingElement" defined in the output model.

This means that TapirLink has limited use with response structures 
that relate instances of different classes, for example multiple 
specimens, each one with multiple identifications. However, in 
these cases sometimes it is possible to use fixed value mappings,
especially when response structures include metadata elements 
(for example "collection code") enclosing all instances of a class.

FEATURES
========

* All TAPIR operations (metadata, capabilities, inventory, search
  and ping). 
* Request encoding can be KVP or XML.
* Inventories on any mapped concepts.
* Searches with any output models involving concepts from mapped 
  schemas.
* Response structures with basicSchemaLanguage.
* Several types of relational databases supported 
  (check http://phplens.com/adodb/supported.databases.html)
* Log only requests can be accepted.
* Complete filter parsing. "Equals" and "like" can be case sensitive 
  or not.
* Max element repetitions and max element levels settings.
* Multiple resources can be exposed from a single TapirLink instance.
* Each resource can map one or more conceptual schemas based on the 
  new DarwinCore pattern or the CNS configuration file format.
* User-friendly Web configuration interface including a UDDI 
  registration form and the possibility to import DiGIR configuration.
* A simple client for testing.
* An LSID authority resolver (see: www/lsid-authority/readme.html)

LIMITATIONS
===========

* Any XML Schema used as a response structure should not include or 
  import other schemas that redeclare the same prefix and associate
  it with a different namespace.

INSTALLATION
============

Please read the INSTALL.txt file for installation instructions.

TROUBLESHOOTING
===============

Try running in your browser the script admin/check.php to see if 
there are any problems with your installation.

If you are having problems with you PHP installation, run the 
scrip admin/info.php to check what are the extensions loaded, 
where is php.ini located, etc.

Activating error logging in your PHP installation is a good idea.
You just need to set the following options in you php.ini file:
log_errors = On
error_log = /some_directory/php.log

If you are getting unexpected blank pages in your browser when running 
scripts, try increasing the memory limit of your PHP instance. Look 
for the option "memory_limit" in your php.ini file. The default value 
in older PHP versions was 8M. Recent versions come with 128M. I suggest
at least 32M. TapirLink consumes more memory if you use it with big and 
complicated response structures or if you set the maximum number of 
element repetitions to a high value.

When you are having problems, it is also a good idea to activate 
debugging. To do this, copy www/localconfig_dist.php to 
www/localconfig.php and enable the line where _DEBUG is set to true. 
Remember to rollback this change when you are in a production environment.

Another PHP setting you might want to adjust is the socket timeout, 
especially if you are going to work with response structures that take 
a long time to load. Look for the "default_socket_timeout" option in the 
php.ini file.

Note: you need to restart your web server everytime you make changes
in php.ini!

If you need to submit any bug report, use the following service 
(category "TapirLink"):

https://sourceforge.net/tracker/?group_id=38190&atid=422311

Since TapirLink accesspoints incorporate the resource code as part of the
URI, some Web Servers may need special configuration to understand that
requests pointing to http://somehost/somepath/tapir.php/someresource
should actually run the tapir.php script. In Apache you can add a line 
like the following one to solve the problem:

AliasMatch (?i)^/somepath/tapir.php/(.*) C:/tapirlinkdir/www/tapir.php

In IIS you will need additional software to produce the same effects 
of rewrite rules. One option is ISAPI Rewrite Lite:

http://www.isapirewrite.com/ 

In this case, these are the steps:

- Install ISAPI Rewrite Lite
- Start IIS
- View Properties of the required Web Site in IIS
- On the ISAPI Filters tab, add a filter (with any suitable name) and
  browse to the installed ISAPI_Rewrite.dll.
- Modify the httpd.ini file in the Rewrite install directory to contain
  the required rewrite rule regular expression, which can be:

RewriteRule ^(.+\/)tapir.php\/([^\?\/]+)[\?\/]?(.*)
$1tapir.php?dsa=$2(?3&$3)

For more information about web server configuration in this case, try 
searching the web for "clean URLs". 

As a last resource, instead of using:

http://somehost/somepath/tapir.php/someresource

You can define your accesspoint as:

http://somehost/somepath/tapir.php?dsa=someresource

DEVELOPER GUIDELINES
====================

Anyone willing to collaborate is welcome. Please contact the author 
about your plans and for SVN access:
renato at cria dot org dot br

TapirLink follows this coding standard:

http://www.dagbladet.no/development/phpcodingstandard/

The following guidelines are also adopted:

* Each class has its own file with the same name.
* Class properties are always manipulated through accessors and mutators.
* Custom library functions that are widely used are also defined inside
  classes and accessed through the :: operator.
* Private and protected methods have an underscore prepended in their name.
* Please give preference to longer and clearer variable, function and class 
  names.
* Do not use TABs for indentation. Use 4 spaces.

It is recommended to subscribe to the following mailing list to get automatic 
notifications about any changes in code:

http://lists.sourceforge.net/mailman/listinfo/digir-cvs

Also remember that TapirLink code should remain compatible with both 
PHP4 and PHP5.


ChangeLog.txt
Release 0.7.3 [Tela Botanica] (2014-07-06)
--------------------------
(GIT revision 17)
	* Added .htaccess to redirect tapir* to www/tapir*
	* Encoded all files to UTF-8
	* Fixed bug with DSN and/or HostName on data source form
	* Fixed bug with dl() function for php 5.3+
	* Fixed bug with @ before include function that don't show the php error messages
	* Fixed bug "Constant _DEBUG already defined" in check.php
	* Added www/localconfig.php à .gitignore

Release 0.7.2 [Tela Botanica] (2013-03-15)
--------------------------
(GIT revision 1)

	* Added a directory "depot" to store xml files (XSD and output models)
	* Added a CNS for Darwin Core terms (2011-10-26) including decisions : Decision-2009-12-07_1 to Decision-2011-10-16_10
	* Fixed XSD bug : renamed dwc:DateTimeISO to dwc:dateTimeISO
	* Fixed XSD bug : deleted last slash in XML namespaces URI
	* Added compatibility with php 5.4 
	* Added .htacces in directories
	* Added index.html (in french) with Twitter Bootstrap template to link all TapirLink tools
	* Fixed bug with UTF8 encoding (we need to believe in user)
	* Fixed bug with the fixed values from mapping when we have a single quotation mark in the texte field

Release 0.7.1 (2010-09-14)
--------------------------
(revision 2027)

	* Resources can now be configured to reject custom output models.
	* Concepts are now automatically mapped by default.
	* New configuration to specify accepted domains for remote file retrieval.
	* New timezone setting.
	* Minor adjustments to avoid warnings in PHP versions >= 5.3.
	* Latest DarwinCore terms added as a possible conceptual schema.
	* New configuration option to allow more efficient counts in databases.

Release 0.7.0 (2009-03-21)
--------------------------
(revision 1988)

	* New admin program to check the database encoding configuration.
	* Check each piece of content that comes from the database 
	  (in search or inventory responses) to avoid wrong charset conversions that
	  may produce invalid XML. Conversion problems could happen when the database
	  charset was incorrectly configured or if the database content somehow
	  didn't follow the expected (real) database encoding.
	* Added support for mixed content elements, allowing TapirLink to be exchange
	  simple TCS (Taxonomic Concept Schema) data.
	* Fixed bug in SQL count caching mechanism when it was disabled.
	* Included basic checks when processing ping request (service configuration, 
	  database connection and runtime errors).
	* When the targetNamespace of the response structure is the RDF namespace 
	  and the envelope is turned off, TapirLink now returns the HTTP Content-type 
	  application/rdf+xml.
	* When a search returns no content and the envelope is turned off, TapirLink 
	  now returns HTTP 204 (No Content).
	* Added support to "IN" operator in KVP filters.
	* New TP_DIAG_LEVEL option can be used to control diagnostics display.
	* Fixed bug when importing DiGIR configuration (resource entity was not being
	  imported and all contacts were accumulating in the host entity across
	  different resources).
	* Replaced XML schema location used in responses with: 
	  http://rs.tdwg.org/tapir/1.0/schema/tdwg_tapir.xsd
	* Added new metadata fields: entity type, region code, country code and zip code.
	* Fixed bug when deserializing bibliographic citation.
	* Added support to the new <rootElement> feature in output model definitions.

Release 0.6.1 (2008-08-01)
--------------------------
(revision 746)

	* Added support for automatic updates.
	* Bugfix: dateLastUpdated environment variable was not working if it
	  was mapped to a field with a datatype other than datetime.
	* Removed unnecessary verification to check if session.save_path is writable
	  when running the configurator.
	* Included caching mechanism for count SQL in inventories and searches. Can
	  be controlled with TP_SQL_COUNT_CACHE_LIFE_SECS.

Release 0.6.0 (2008-04-30)
--------------------------
(revision 660)

	* Bugfix: When different configuration interfaces from different 
	  TapirLink installations were being accessed through the same browser,
	  session caching was not making any distinction between them.
	* Users now need to explicitly choose the conceptual schema format when
	  trying to load an additional conceptual schema in the configuration 
	  interface (mapping form).
	* When loading concepts from a CNS configuration file (old text format), if the 
	  concept_source alias was not specified then the last concept_source
	  will be loaded.
	* Bugfix: Resuming the inclusion of new resources was probably not working 
	  correctly in previous versions.
	* Configuration interface can now make use of debug logging.
	* Bugfix: KVP filters were not working properly on PHP 4.x when they had
	  nested parenthesis.
	* Many adjustments in the statistics module.
	* Fixed bug that was affecting KVP requests when parameter names were in uppercase.
	* Included missing "charset=utf-8" in the HTTP header.
	* Flush cache after any change in the local mapping.
	* Resources are not serialized in session anymore when running the web service.
	* Bugfix: Empty tagname parameters (inventory operation) were not being ignored.
	* Bugfix: Output models were failing to produce results when the indexing
	  element had a simple type.
	* Bugfix: Searches were failing when the parameter "limit" was not specified
	  (default was being set to zero instead of null).
	* Follow up protocol update: Filter condition will now be ignored when parameter
	  is not passed. When a concept is not locally mapped, the error will now only be
	  raised if there's no missing parameter.
	* Bugfix: KVP filters were not working if they contained any of the following
	  operators: lessthan, lessthanorequals, greaterthan, greaterthanorequals.
	* Bugfix: Search XML requests referencing external templates were not working.
	* Updated local version of ADODB to 4.98.
	* Changed DiGIR configuration import procedure to overcome an XML parsing problem

www/tapir.php
<?php
/**
 * $Id: tapir.php 1997 2009-09-07 22:45:02Z rdg $
 * 
 * LICENSE INFORMATION
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details:
 * 
 * http://www.gnu.org/copyleft/gpl.html
 * 
 * 
 * @author Renato De Giovanni <renato [at] cria . org . br>
 * @author Dave Vieglais (Biodiversity Research Center, University of Kansas)
 * 
 * ACKNOWLEDGEMENTS
 * 
 * TapirLink has been generously funded by the Biodiversity 
 * Information Standards, TDWG, with resources from the Gordon and 
 * Betty Moore Foundation. The Global Biodiversity Information 
 * Facility, GBIF, has also been a major supporter of the TAPIR 
 * initiative since its very beginning and also collaborated to 
 * test this software.
 *
 * TapirLink was based on the DiGIR PHP provider, which was 
 * originally developed by Dave Vieglais from the Biodiversity 
 * Research Center, University of Kansas. The software is now being 
 * distributed under the GPL with permission from its original
 * author. The original copyright information is also 
 * reproduced below.
 * 
 * --------------------------------
 * 
 * Copyright (c) 2002 The University of Kansas Natural History 
 * Museum and Biodiversity Research Center. All rights reserved.
 * 
 * Permission is hereby granted, free of charge, to any person 
 * obtaining a copy of this software and associated documentation 
 * files (the "Software"), to deal with the Software without 
 * restriction, including without limitation the rights to use, 
 * copy, modify, merge, publish, distribute, sublicense, and/or 
 * sell copies of the Software, and to permit persons to whom the 
 * Software is furnished to do so, subject to the following 
 * conditions:
 * 
 * - Redistributions of source code must retain the above copyright 
 * notice, this list of conditions and the following disclaimers.
 * 
 * - Redistributions in binary form must reproduce the above 
 * copyright notice, this list of conditions and the following 
 * disclaimers in the documentation and/or other materials provided 
 * with the distribution.
 * 
 * - Neither the names of The University of Kansas Natural History 
 * Museum and Biodiversity Research Center, The University of Kansas 
 * at Lawrence, nor the names of its contributors may be used to 
 * endorse or promote products derived from this Software without 
 * specific prior written permission.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES 
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND 
 * NONINFRINGEMENT. IN NO EVENT SHALL THE CONTRIBUTORS OR COPYRIGHT 
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, 
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
 * DEALINGS IN THE SOFTWARE.
 */

// Definition of this constant can be used to indicate that
// the Web Service (tapir.php) was called
define( 'TP_RUNNING_TAPIR', 1 );

require_once('tapir_globals.php');
require_once('TpUtils.php');
require_once('TpServiceUtils.php');
require_once('TpDiagnostics.php');
require_once('TpResources.php');
require_once('TpRequest.php');
require_once('TpResponse.php');

// Register a new error handler
$old_error_handler = set_error_handler( 'TapirErrorHandler' );

// Avoid HTML in errors
ini_set( 'html_errors', false );

// Store timestamp for profiling
define( 'INITIAL_TIMESTAMP', TpUtils::MicrotimeFloat() );

// Logs
TpUtils::InitializeLogs();

// Main logger
global $g_log;

// Debug logger
global $g_dlog;

$g_dlog->debug( '[Starting debug log]' );
$g_dlog->debug( 'TapirLink version: '.TP_VERSION.' (revision '.TP_REVISION.')' );

$request_uri = ( isset( $_SERVER['REQUEST_URI'] ) ) ? $_SERVER['REQUEST_URI'] : 'null';

$g_dlog->debug( 'Request URI: '.$request_uri );
$g_dlog->debug( 'Include path: '.ini_get( 'include_path' ) );

// Set the maximum script run time
// Note: if php set safe mode is on then this can't be set from here, 
// and will need to be set in php.ini file
set_time_limit( TP_MAX_RUNTIME );

// Debugging
if ( ! defined( '_DEBUG' ) )
{
    // Load the value from a request variable called "debug", default is false
    // for security reasons
    if ( TP_ALLOW_DEBUG )
    {
        define( '_DEBUG', (bool)TpUtils::GetVar( 'debug', false ) );
    }
    else
    {
        define( '_DEBUG', false );
    }
}

// Instantiate request object
$request = new TpRequest();

// If no resource was specified in the request URI, then dump some help information
if ( ! $request->ExtractResourceCode() )
{
   include_once( 'index.php' );
   die();
}

// Get resource code and check if it's valid
$resource_code = $request->GetResourceCode();

$r_resources =& TpResources::GetInstance();

$raise_errors = false;
$r_resource =& $r_resources->GetResource( $resource_code, $raise_errors );

if ( $r_resource == null )
{
    $response = new TpResponse( $request );
    $response->ReturnError( 'Resource "'.$resource_code.'" not found.' );
    die();
}

$r_resources->SetCurrentResourceCode( $resource_code );

if ( $r_resource->GetStatus() != 'active' )
{
    $response = new TpResponse( $request );
    $response->ReturnError( 'Resource "'.$resource_code.'" is not active.' );
    die();
}

// Check PHP version
$current_version = phpversion();

if ( version_compare( $current_version, '5.0', '<' ) > 0 )
{
    if ( version_compare( $current_version, TP_MIN_PHP_VERSION, '<' ) > 0 )
    {
        $msg = 'PHP Version '.TP_MIN_PHP_VERSION.' or later required. '.
               'Some features may not be available. Detected version '.$current_version;
        TpDiagnostics::Append( DC_VERSION_MISMATCH, $msg, DIAG_WARN );
    }
}
else if ( version_compare( $current_version, '6.0', '<' ) > 0 )
{
    if ( version_compare( $current_version, '5.0.3', '<' ) > 0 )
    {
        // Avoid bug in "xml_set_start_namespace_decl_handler"
        $msg = 'Provider error: Unsupported PHP version ('.$current_version.'). To '.
               'use PHP5 it is necessary to have at least version 5.0.3';

        $response = new TpResponse( $request );
        $response->ReturnError( $msg );
        die();
    }
}

// Get parameters
if ( ! $request->InitializeParameters() or 
     TpDiagnostics::Count( array( DIAG_ERROR, DIAG_FATAL ) ) )
{
    $response = new TpResponse( $request );
    $response->ReturnError( 'Failed to process request' );
    die();
}

// By default, assume that the database encoding cannot be detected 
// by the mb_detect_encoding PHP function
global $g_encoding_can_be_detected;

$operation = $request->GetOperation();

if ( $operation == 'ping' )
{
    require_once('TpPingResponse.php');

    $response = new TpPingResponse( $request );
}
else if ( $operation == 'capabilities' )
{
    require_once('TpCapabilitiesResponse.php');

    $response = new TpCapabilitiesResponse( $request );
}
else if ( $operation == 'metadata' )
{
    require_once('TpMetadataResponse.php');

    $response = new TpMetadataResponse( $request );
}
else if ( $operation == 'inventory' )
{
    require_once('TpInventoryResponse.php');

    $response = new TpInventoryResponse( $request );
}
else if ( $operation == 'search' )
{
    require_once('TpSearchResponse.php');

    $response = new TpSearchResponse( $request );
}
else
{
    // Unknown operation 
    $response = new TpResponse( $request );
    $response->ReturnError( 'Unknown operation "'.$operation.'"' );
    die();
}

$response->Process();

exit();

?>
